Add a reshape_copy operator.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88361
* #88063
* __->__ #88314
* #88329

The semantics is "as if" you did a reshape, but it always copied
even if the input was directly view'able.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add a reshape_copy operator.

The semantics is "as if" you did a reshape, but it always copied
even if the input was directly view'able.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2984,3 +2984,7 @@
 - name: special_spherical_bessel_j0(Tensor x) -> Tensor
   x: non_differentiable
+
+- name: reshape_copy(Tensor self, SymInt[] size) -> Tensor
```
Would it work to make it a `copy: bool` kwarg like we did for `.to()` https://pytorch.org/docs/master/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to ?

ezyang replied:
Even if I did add the kwarg, I would still have to add an extra operator, similar to how we have `_to_copy`

albanD reviewed (approved):
Just a question about consistency with .to()
But if you need an actual different function and the flag is not enough, the PR sounds good.

ezyang committed:
Update on "Add a reshape_copy operator."


The semantics is "as if" you did a reshape, but it always copied
even if the input was directly view'able.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add a reshape_copy operator."


The semantics is "as if" you did a reshape, but it always copied
even if the input was directly view'able.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1574,6 +1574,23 @@ Tensor reshape_symint(const Tensor& self, c10::SymIntArrayRef proposed_shape) {
+    TORCH_CHECK(0, "_reshape_copy not implemented for mkldnn tesnors");
+  }
+
+  if (self.is_contiguous()) {
```
do you need this conditional, you are making a clone anyway?

ezyang replied:
The guts here don't really matter, the inside of this function is never traced.

ngimel replied:
Unsafe view is safe really, it does check the sizes, it just doesn't set up autograd tracking that you don't need here.

ezyang replied:
ok, I will update this

ezyang committed:
Update on "Add a reshape_copy operator."


The semantics is "as if" you did a reshape, but it always copied
even if the input was directly view'able.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]
